Generalize `geometry.Line`

siefkenj opened the pull request:
This generalizes the functions in `line.py` to work in dimensions higher than 2 where possible.  It is a step along the way to unifying `Line` and `Line3D` in a similar way to the unification of `Point` and `Point3D`.  Incidentally, the generalized routines are shorter, but relying on the `Matrix` object seems to make them slower.  If speed is an issue, optimized routines for particular dimensions can be added after `Line` and `Line3D` are unified.

siefkenj committed:
Generalize geometry.LinearEntity for n-d

	Many of the routines in line.py can
	be generalized to work with n-dimensional
	lines.  Where possible these functions have been
	generalized to ease the potential merger of Line
	and Line3D.

siefkenj committed:
Unification of LinearEntity and LinearEntity3D

siefkenj committed:
Fixed import of Line3D

siefkenj committed:
Removed erroneous Exception

siefkenj committed:
Fixed Pretty-printing of GeometryEntity

siefkenj committed:
Documentation fixes for migrated LinearEntity3D

siefkenj committed:
Avoid excessive conversion of `Point`s

	Many routines in the Geometry module call
	`Point(a)` on their argument to turn iterables
	into points.  If `a` is already a `Point`, this
	is uneeded.  This patch adds `Point.pointify`,
	which will check if the argument is already a point
	and only convert it if need be.  It also has optional
	arguments to ensure something is a `GeometryEntity`.

siefkenj committed:
Merge branch 'generalize' of github.com:siefkenj/sympy into generalize

siefkenj committed:
Added compatability for useing Point2D with 3D objects

	* Previous behavior allwed passing `Point2D` objects
	into 3D objects.  A zero would be added to the end of them.
	This behavior is now allowed again.

	* The numerical approximations used for an ellipse are now
	numerically checked instead of tested for perfect equality.

siefkenj committed:
Added tests for parameter types

	If a Point is allowed as an argument for a method
	in the geometry module, an iterable should be allowed aswell.
	This adds tests to `point.py` and `line.py` to ensure
	compatability.

siefkenj committed:
Fix assignment to `self`

siefkenj committed:
Generalized concyclic

siefkenj committed:
Changed doc string

siefkenj committed:
Reverted docstring

siefkenj committed:
Revert "Fix assignment to `self`"

This reverts commit 35cec38031f01ffdeab13fc4d46867e743c1388b.

siefkenj committed:
Trigger travis rebuild...

siefkenj committed:
Finished revert

siefkenj committed:
Revert "Finished revert"

This reverts commit 19c707e6f7e4cb5e36ec5e7df28dc95aa665871b.

siefkenj committed:
Revert "Trigger travis rebuild..."

This reverts commit 1a62444fe24c6d498177c4937fedc647f9b17d33.

siefkenj committed:
Revert "Revert "Fix assignment to `self`""

This reverts commit f0472008a9e3532f4eba0235471709b5a7fb0a04.

siefkenj committed:
Revert "Reverted docstring"

This reverts commit 9f6930505cbf0b210273ac3113f661e78415eff4.

siefkenj committed:
Revert "Changed doc string"

This reverts commit 417dcbda44c627e5d099fca6dc589cfb104a9967.

siefkenj committed:
Revert "Generalized concyclic"

This reverts commit f1da2200e3f173a32293031bce00eb6f8a7e73b8.

siefkenj commented:
@smichr Could you take a look at this?  I will resolve the new conflicts as soon as I can.

siefkenj committed:
Merge remote-tracking branch 'upstream/master' into generalize

siefkenj committed:
Finished merge

siefkenj committed:
Fixed intersection bug and concurrency tests

	There was an error when finding the intersection of
	a line with a `FiniteSet` which resulted in an `Intersection`
	operation failing with `PYTHONHASHSEED=1610241886`  This has
	been fixed. Aswell, the unit tests for concurrency of 3D lines
	were incorrect.

siefkenj committed:
Cleaned unused code

smichr commented:
Didn't see the ping...I can look at this over the coming week.

smichr commented:
The way work is going right now I can see this might take longer than this week. If someone else wants to review it, that would be fine with me.

siefkenj commented:
@smichr do you have time to take a look at this PR yet?

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -835,23 +835,23 @@ def normal_lines(self, p, prec=None):
         >>> e.normal_lines((3, 3), prec=2)
-        [Line(Point2D(-38/47, -85/31), Point2D(9/47, -21/17)),
-        Line(Point2D(19/13, -43/21), Point2D(32/13, -8/3))]
+        [Line2D(Point2D(-0.81, -2.7), Point2D(9/47, -21/17)),
```
Since the evaluated points are Points when the Line is created their decimal approximations are not recomputed to fractions. I think this can be fixed by wrapping the numerically approximated point in `Point`: `points = [Point(pt.n(prec)) for pt in points]` at line 895.

smichr commented on sympy/geometry/line.py:
```diff
@@ -36,16 +46,14 @@ class Undecidable(ValueError):
 
     p1
     p2
-    coefficients
```
Why are these deleted? Beside `p1, p2, and points` there are also `ambient_dimension, coefficients, direction and length`.

siefkenj replied:
I see.  coefficients should be moved to `Line2D` since it's only relevant there.

smichr commented on sympy/geometry/line.py:
```diff
@@ -270,19 +299,25 @@ def is_perpendicular(l1, l2):
+        if not isinstance(l1, LinearEntity) and not isinstance(l2, LinearEntity):
+            raise TypeError('Must pass only LinearEntity objects')
+
+        return l1.direction.dot(l2.direction) == S.Zero
```
Isn't this made weaker by not including a call to `simplify` -- `dot` doesn't do simplification? I wonder if returning an `Eq` which might be unevaluated might be better. One would have to see how this is handled elsewhere.

siefkenj replied:
Nice idea.  I'll run it through the test suite.

smichr commented on sympy/geometry/line.py:
```diff
@@ -394,19 +443,20 @@ def perpendicular_line(self, p):
-        else:
-            p2 = Point(p.x - d2, p.y + d1)
-            return Line(p, p2)
+        p = Point.pointify(p, dimension=self.ambient_dimension)
```
[that cleaned up nicely]

smichr commented on sympy/geometry/line.py:
```diff
@@ -447,26 +497,25 @@ def perpendicular_segment(self, p):
-            dy = m*dx
-            p2 = (p.x + dx, y + dy)
+        l = self.perpendicular_line(p)
+        p2, = Intersection(Line(self.p1, self.p2), l)
```
Delete the comma after `p2`?

siefkenj replied:
`p2, = Intersection(Line(self.p1, self.p2), l)` is a sneaky way of unpacking an iterable with one element.  Since `Intersection` is a set, I cannot do `Intersection(Line(self.p1, self.p2), l)[0]`.  Should I add a comment explaining it is an unpacking, or should I replace it with `p2 = list(Intersection(Line(self.p1, self.p2), l))[0]`?

smichr replied:
It is subtle enough that I think it warrants a comment. It should also probably include a note; `pop` is nice for such cases, too:

```python
p2 = Intersection(Line(self.p1, self.p2), l).pop() # there can only be one point of intersection btwn lines
```

siefkenj replied:
unfortunately, `FiniteSet` doesn't have an attribute `pop`...

smichr commented on sympy/geometry/line.py:
```diff
@@ -601,43 +604,45 @@ def projection(self, o):
+            projected = Intersection(self, projected)
+            # if we happen to have intersected in only a point, return that
+            if projected.is_FiniteSet and len(projected) == 1:
+                a, = projected
```
There's that comma again...I don't recognize what its function is.

smichr commented on sympy/geometry/line.py:
```diff
@@ -663,141 +668,134 @@ def intersection(self, o):
-            return []
-
-        return o.intersection(self)
+        return other.intersection(self)
```
This is organized well and is clear. I leave the verification to the test suite.

smichr replied:
:end: reviewed to here

smichr commented on sympy/geometry/point.py:
```diff
@@ -172,28 +176,31 @@ def is_collinear(*args):
-            raise TypeError('Must pass only Point objects')
+        # XXX: the current workaround for issue #11238 relies on the points
+        # being in a specific order, so don't let `set` change the order
+        #args = list(set(args))
```
It would be better to leave this as it is and XFAIL the test. Also, does #11554 fix this? Hopefully so, because when this PR is merged with the current master the following fails:

```python
x = 8*tan(13*pi/45)/(tan(13*pi/45) + sqrt(3))
y = (-8*sqrt(3)*tan(13*pi/45)**2 + 24*tan(13*pi/45))/ \
    (-3 + tan(13*pi/45)**2)
assert Line(Point(0, 0), Point(1, -sqrt(3))).contains(Point(x, y)) is True
```

Some orderings of the points works and some don't. I suspect it has to do with the fact that the ratio of `y/x` simplifies to `-sqrt(3)` but the coordinates themselves don't simplify enough for the routine to succeed.

The code you deleted -- using `is_scalar_multiple` works, however. Any reason not to keep it?

smichr replied:
Never mind -- I see that only 4 of the 6 permutations of point orderings work. But the docstring should be updated to reflect that slopes are not being checked.

siefkenj replied:
I think #11554 does fix this.  I tried to update the docstring before and got really frustrated with sphinx errors.  I'll try again.

smichr commented on sympy/geometry/line.py:
```diff
@@ -874,36 +877,19 @@ def random_point(self):
+            upper = floor(self.length)
+        t = randint(lower, upper)
+
+        return self.direction*t/abs(self.direction) + self.p1
```
nice!

smichr commented on sympy/geometry/point.py:
```diff
@@ -172,28 +176,31 @@ def is_collinear(*args):
-    def is_scalar_multiple(p1, p2):
+    def is_scalar_multiple(self, p2):
         """Returns whether `p1` and `p2` are scalar multiples
         of eachother.
```
Could you find and replace `eachother` -> `each other`?

smichr commented on sympy/geometry/line.py:
```diff
@@ -999,33 +1000,19 @@ class Line(LinearEntity):
+            ambient_dim = len(p1.p1)
         else:
-            raise ValueError('A 2nd Point or keyword "slope" must be used.')
+            if iterable(p1) and not hasattr(p1, '__len__'):
```
I think you might want to use `is_sequence` instead of `iterable`...probably throughout the PR. Please double check.

siefkenj replied:
I did this to be efficient.  Perhaps it is confusing.  `x=[1, 2, 3]` will satisfy `is_sequence(x) == True`, but it will not make it through the if clause since `hasattr(x, '__len__') == True`.  Thus, I avoid an unnecessary call to `tuple(x)`.  Should I be doing a `try` `except` approach instead?

smichr commented on sympy/geometry/line.py:
```diff
@@ -1057,50 +1044,9 @@ def plot_interval(self, parameter='t'):
-        return a*x + b*y + c
-
-    def contains(self, o):
+    def contains(self, other):
```
:end: reviewed to here

smichr commented on sympy/geometry/point.py:
```diff
@@ -234,6 +256,95 @@ def ambient_dimension(self):
+
+        # make sure we're genuinely points
+        # and translate to every point to the origin
+        # XXX: issue #11238 we need to pre-simplify otherwise
```
preceding lines:

# make sure all points are Points
# and translate all points to the origin

Regarding the 11238 hacks: could you add a test in the geometry suite which, if it passes, directs someone to go back through geometry and make the appropriate modifications? In case that isn't you, it would be good to include the post-fix line(s) as one or more comments and indicate which line(s) they are replacing.

smichr commented on sympy/geometry/point.py:
```diff
@@ -234,6 +256,95 @@ def ambient_dimension(self):
+        # (which excludes things like FiniteSets) can be turned
+        # into a point.  This check can be made more general, but for now
+        # exclude known bad objects
+        if iterable(a) and not isinstance(a, (FiniteSet, set, dict)):
```
I suspect `if is_sequence(a):` would work.

siefkenj replied:
I avoided `is_sequence` so that `Point(x**2 for x in [1,2,3])` would work.  That is, `g=(x**2 for x in [1,2,3])` passes the `pointify` tests, but `is_sequence(g) == False`.

smichr replied:
I like the generator instantiation. 

I think `is_sequence(a, include=GeneratorType)` will fit the bill.

siefkenj replied:
I don't know how generous or non-generous we'd like to be, but `is_sequence(iter([1,2,3]), include=GeneratorType) == False`.  We could decide to exclude iterators and only allow sequences or generators, or we could leave the check quite general.  I think passing a long list of exceptions to is_sequence is kindof ugly...(the code for is_sequence isn't terribly complicated.  We'd basically be asking is_sequence to ignore most of its functionality).

smichr replied:
Let's insist on literal sequences...you can still use a generator but you have to unpack it yourself which is easy:

```python
>>> def g(*a):
...  print a
...
>>> g(*(i**2 for i in range(3)))
(0, 1, 4)
>>> g(*iter([1,2,3]))
(1, 2, 3)
```

Also, if you could look at the `__new__` routines for `Point` and `pointify` and see if they can be a bit less convoluted (at least that's how it seems to me right now) that would be good. I'll push some ideas to a `generalize` branch if you want to take a look.

siefkenj replied:
Sounds good to me.  I'll have a look at this tomorrow.

siefkenj replied:
Do you mean `Point2D.__new__` and `Point3D.__new__`?  `Point.__new__` doesn't seem strange to me...though I have no idea why we prevent imaginary coordinates.

smichr commented on sympy/geometry/point.py:
```diff
@@ -394,19 +506,22 @@ def intersection(self, o):
-        p2 = Point(p2)
-        return Add(*[a*b for a,b in zip(self, p2)])
+        p2 = Point.pointify(p2, dimension=self.ambient_dimension)
+        # sum could return native python floats, we'd like to ensure the
```
Now that you are using pointify I don't think this "native python floats" situation applies so you can leave the `S()` off.

smichr replied:
If there isn't a test of this, one should be added, e.g., one where `p2 = (0.1, 0.2)`

smichr commented on sympy/geometry/point.py:
```diff
@@ -442,32 +557,44 @@ def __add__(self, other):
+            coords = [simplify(a + b) for a, b in zip(self, other)]
+            if len(coords) != self.ambient_dimension:
+                raise ValueError("Points must have the same number of dimensions")
+            if all(isinstance(c, Number) for c in coords):
```
here and below, I'm not sure this is a good idea...can you explain the rationale of `Point(.1,.2) -> Point(1/10, 1/5)` but `Point(1,2)*0.1 -> Point(0.1, 0.2)`?

siefkenj replied:
I don't know the history of `Point` turning all its arguments into exact types, so I didn't mess with that, but I tried to make it avoid coercing types when it wasn't needed.  (There are some bug's filed about poor numeric performance of `Point`).  I think the best behaviour would be for `Point(.1,.2) -> Point(.1,.2)`.  I suspect this would break a lot of the library though, since things like distance calculations, etc. would then be floats and I think a lot of the library has been programmed with the assumption of exactness...

smichr replied:
OK, I'm in agreement. When performing a calculation like `S(1)/3.0` one gets a Float so when multiplying a Point by a float I guess one shouldn't be surprised that it makes the Point inexact.

smichr commented on sympy/geometry/point.py:
```diff
@@ -442,32 +557,44 @@ def __add__(self, other):
         factor = sympify(factor)
-        return Point([simplify(x*factor) for x in self.args])
+        coords = [simplify(x*factor) for x in self.args]
+        if all(isinstance(c, Number) for c in coords):
```
see above

smichr commented on sympy/geometry/point.py:
```diff
@@ -442,32 +557,44 @@ def __add__(self, other):
         """Negate the point."""
-        return Point([-x for x in self.args])
+        coords = [-x for x in self.args]
+        if all(isinstance(c, Number) for c in coords):
```
see above

smichr commented on sympy/geometry/tests/test_line.py:
```diff
@@ -168,12 +169,11 @@ def test_line_geom():
     assert s1.intersection(Line(p6, p9)) == []
     s3 = Segment(Point(0.25, 0.25), Point(0.5, 0.5))
-    assert s1.intersection(s3) == [s1]
+    assert s1.intersection(s3) == [s3]
```
hmm...glad that got fixed

smichr commented on sympy/geometry/tests/test_line.py:
```diff
@@ -192,7 +192,7 @@ def test_line_geom():
-    a, b = symbols('a,b')
+    a, b = symbols('a,b', real=True)
     s = Segment((0, a), (0, b))
     assert Point(0, (a + b)/2) in s
```
I don't think we want this to fail when vanilla (non-real) symbols are used. I didn't catch the change that made it so -- is it possible to relax the test so we don't get Undecided?

```python
>>> var('a b')
(a, b)
>>> s = Segment((0, a), (0, b))
>>> Point(0, (a + b)/2) in s
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\geometry\line.py", line 914, in __contains__
    result = self.contains(other)
  File "sympy\geometry\line.py", line 1653, in contains
    raise Undecidable("Cannot determine if {} is in {}".format(other, self))
sympy.geometry.line.Undecidable: Cannot determine if Point2D(0, a/2 + b/2) is in
 Segment2D(Point2D(0, a), Point2D(0, b))
```

siefkenj replied:
Good thinking.  I've replaced my containment check with one that relies on the triangle inequality instead of the dot product.

smichr commented on sympy/geometry/tests/test_line.py:
```diff
@@ -220,10 +220,10 @@ def test_line_geom():
     l = Line((0, 5), slope=m)
     p = Point(2, 3)
-    assert l.distance(p) == 2*abs(m + 1)/sqrt(m**2 + 1)
+    assert simplify(l.distance(p) - 2*abs(m + 1)/sqrt(m**2 + 1)).is_zero
```
Did the distance routine lose a simplification step? The expression obtained without simplification is `sqrt((3 - (3*m**2 + 2*m + 5)/(m**2 + 1))**2 + (-(-2*m + 2)/(m**2 + 1) + 2)**2)`.

siefkenj replied:
In general, I removed as many calls to `simplify` as I could.  I guess this test should really use `.equals`

smichr commented on sympy/geometry/tests/test_line.py:
```diff
@@ -393,7 +393,7 @@ def test_line3d():
     assert line3.is_parallel(line4) is False
     assert Line3D((0, 1, 2), (0, 2, 3)).intersection(
-        Line3D((0, 1, 2), (0, 1, 1))) == []
+        Line3D((0, 1, 2), (0, 1, 1))) == [Point3D(0,1,2)]
```
yes

smichr commented:
OK, I've made it through. 

- Overall,  I see that this has benefited from mathematical reformulation.

- I didn't like `pointify` at the start, but it grew on me. I think it's the right thing to do.

- I left a few concerns which you have already started to comment on. I will try to check back in this week.

siefkenj committed:
Merge remote-tracking branch 'upstream/master' into generalize

siefkenj commented:
I have made several adjustments as per your comments.  When #11554 is merged, I will push them (otherwise a bunch of tests will just fail).

smichr commented:
#11554 is in.

siefkenj committed:
Merge remote-tracking branch 'upstream/master' into generalize

siefkenj committed:
Code review changes

	* Fixed documentation
	* made LinearEntity.contains work for symbols without declaring
	them as real

siefkenj committed:
Removed workarounds for issue #11238

smichr commented on sympy/geometry/tests/test_point.py:
```diff
@@ -260,3 +260,61 @@ def test_concyclic_doctest_bug():
+    for func in test_double:
+        for p in doubles2d:
+            getattr(p3d, func)(*p)
+    p3d.are_coplanar((1,2,3), [1,2,2], RepeatableGenerator((3,3,2)))
```
this line has the wrong syntax and problem should be removed as it is not testing anything

siefkenj replied:
Can you explain a bit more of what you mean?  That line is supposed to be testing that `are_coplanar` accepts sequence-like inputs.

smichr replied:
I have made several changes (including this one) which I will send as a friendly PR tomorrow morning.

siefkenj commented:
What's the status?  Are there things I need to address?

smichr committed:
pointify modified

smichr committed:
pointify removed; Point modified; other changes

* Point now allows one to specify the desired dimension.
* Point will return Point instances unchanged except for dimension change
  (if needed)
* the ellipse docstring was modified.
* moved Undecidable to utilities.misc
* normalize_dimensions was added to assure that one is working with Points of
  matching dimensions
* docstring added to direction
* segment intersection works with differing dimensions; test added for that and
  other similar routines
* remove use of set() to remove duplicates; use uniq instead which makes
  routines canonical in behavior since it preserves order
* private _size property added to give the nominal dimension of a point
* are_coplanar added to Point; docstring moved there
* are_coplanar added to Point2D to return True for 2-D points
    maybe bad because then the general mutation of dimensions doesn't
    apply
* signature for is_concyclic changed to reflect that they are methods, not
  staticmethods: self added as first arg. raise error in Point's method if the
  dimension is not * supported to avoid recursion error.
* added docstring to orthogonal_direction
* unit method added to Point to normalize the coordinates of a point so it is in
  the same direction but a distance of 1 from the origin
* correct logic of Segment.perpendicular_line to not return None

smichr committed:
review modifications

siefkenj committed:
Merge pull request #1 from smichr/generalize_cps

wip

siefkenj committed:
Mutating dimension gives warning--dimension never mutated down

This patch also generalizes `is_concyclic`.

smichr commented on sympy/geometry/point.py:
```diff
@@ -245,18 +214,16 @@ def are_coplanar(*points):
-
-        Test whether or not a set of points is collinear. Returns True if
-        the set of points are collinear and False otherwise.
+        """Retursn `True` if there exists a line
```
Retursn -> Returns

smichr replied:
Retursn -> Returns

smichr commented on sympy/geometry/point.py:
```diff
@@ -245,18 +214,16 @@ def are_coplanar(*points):
-        Test whether or not a set of points is collinear. Returns True if
-        the set of points are collinear and False otherwise.
+        """Retursn `True` if there exists a line
+        that contains `self` and `points`.  Returns `False` otherwise.
```
You should add the "trivially true" comment here, too.

smichr commented on sympy/geometry/point.py:
```diff
@@ -127,93 +149,40 @@ def __new__(cls, *args, **kwargs):
-            points = [Point(a, dim=dim, warn=warn) for a in points]
-        return points
+    @classmethod
+    def _normalize_dimension(cls, *args, **kwargs):
```
I preferred to require that the args be Points since the calling routine knows better what is and isn't a Point. The way it is now, if dim is not given then two additional Point calls are being made. It was my impression that Point is rather slow. Perhaps with the mods the speed for handling Point instances is better and this is no longer a significant issue?

smichr replied:
I preferred to require that the args be Points since the calling routine knows better what is and isn't a Point. The way it is now, if dim is not given then two additional Point calls are being made. It was my impression that Point is rather slow. Perhaps with the mods the speed for handling Point instances is better and this is no longer a significant issue?

siefkenj replied:
Having all coercions to `Point` in one place would ensure that the behaviour is uniform.  The code is likely slightly slower than the old code with `pointify`, but I think it's significantly faster than current master.  Calling `Point(Point(..))` should have almost no overhead. `_normalize_dimension` could be made to avoid the excess calls, but I couldn't think of a way to do it without making the function uglier.

smichr commented on sympy/geometry/point.py:
```diff
@@ -404,54 +371,55 @@ def ambient_dimension(self):
 
         # make sure we're genuinely points
         # and translate every point to the origin
-        points = Point.normalize_dimensions(*[Point(i) for i in points])
```
If you don't normalize you can get a Matrix error with different length rows: `Point.affine_rank(Point(1,2),Point(1,2,3))`

smichr replied:
If you don't normalize you can get a Matrix error with different length rows: `Point.affine_rank(Point(1,2),Point(1,2,3))`

smichr commented:
Sorry about multiple comments...my browser seems to be having a problem (I have to enter the Preview tab to enter the comment!) and/or I am not understanding how the new review process works.

Note to self: What's the difference between staticmethod and classmethod?

smichr commented:
Sorry about the conflicts with the new distance calculation. That's just a temporary fix -- go ahead and replace with what we have here.

siefkenj committed:
Merge remote-tracking branch 'upstream/master' into generalize

smichr commented:
Since all other routines convert non-point args to Point with `evaluate=True` we should think more about whether `Point(1,1)+(.1,.2)` should be `Point(1.1, 1.2)` or `Point(11/10, 6/5)`. The former is consistent with other methods of Point while the latter is more consistent with Number operations.

siefkenj commented:
I think maybe `Point(1,1)+(.1,.2)` should raise an exception and we should only auto-convert sequences when they are explicitly passed as arguments to methods.

siefkenj committed:
Force warnings to display in unit tests

smichr commented:
> should raise an exception and we should only auto-convert sequences when they are explicitly passed as arguments to methods

OK. `translate` is a method that will do addition. Perhaps we should add a `skew` method that does multiplication: `Point(1,2,3).skew(4) -> Point(4, 2, 3); Point(1,2,3).skew(4, 4, 4) -> Point(4, 8, 12); `.

siefkenj commented:
We already have `scale` which does the exact same thing.

smichr commented:
> `scale`

I even looked at that.... OK, nothing new needed. I would leave the current behavior as it is wrt adding non-Points, then. At some point, however, the scale and translate methods should be methods of the general Point rather than only methods of Point2D and Point 3D, perhaps with scale having the signature `(*args, **kwargs)` where kwargs recognizes `pt` as giving the point from which the scaling happens.

smichr commented:
Are you going to address the issues that were raised in the most recent review?

smichr committed:
minor geom modifications

smichr committed:
modify ambient_dimension docstring

siefkenj commented:
Most of these issues are fixed in your pull request.  I will merge it as soon as the `on_morph` options make a little more English sense.

smichr committed:
on_morph must be warn, ignore, or error

siefkenj committed:
Merge pull request #2 from smichr/generalize

minor mods

smichr reviewed (approved):
I'll commit tomorrow unless there is more that you wanted to do. Thanks for your work on this.

siefkenj commented:
Great.  Thank you for the help.  Nothing else I was hoping for in this PR. (I was going to alphabetize the methods, but that is best for a different PR).

siefkenj committed:
Fixed docstring

siefkenj committed:
Changed docstring

siefkenj commented:
Okay...I'm really confused what is wrong with the docstrings to make sphinx upset.  Do you see the error?

siefkenj committed:
removed doc string

smichr commented:
> Okay...I'm really confused what is wrong with the docstrings to make sphinx upset.  Do you see the error?

No. I twiddled it, too. I don't think we need the `:math:`. You might try to delete the whole thing in an editor and retype it in case there is a gremlin in there. I *have* seen that happen before. Maybe @asmeurer has seen this before.

siefkenj commented:
I tried deleting the whole docstring and the error has moved to another docstring...

smichr commented:
> I tried deleting the whole docstring

OK, that's good news. Did it move further down in the file? If so, that means the "start" code that it is recognizing that doesn't have an end code so by doing a binary search on the file contents you should be able to figure out what the start code is. e.g. delete file from halfway above `ambient_dimension` to the end and see if the error shows up. If so, delete the half above that, etc...

siefkenj commented:
@smichr I am not able to run the sphinx test on my machine.  Are you able to do this bisection?

jksuom commented on sympy/geometry/point.py:
```diff
@@ -77,82 +107,160 @@ class Point(GeometryEntity):
+    @classmethod
+    def are_coplanar(cls, *points):
+        """Return True if there exists a plane in which all the points
+        lie.  A trivial True value is returned if `len(points) < 3 or
```
len(points) < 3 has only one backquote.

siefkenj replied:
Thanks for finding this!

siefkenj committed:
Fixed docstring

smichr commented:
OK! It's in.
